tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerSession.java

```diff
@@ -364,174 +353,38 @@ ResourceCleanupStrategy resourceCleanupStrategy() {
-     *
-     * <p>Clearing a NativeReference does not invoke {@link #delete()}, thus won't release the
-     * native resources it refers to. It can be used when passing the ownership of those resources
-     * to another object.
-     *
-     * <p>If native resources needs to be deleted as well, call {@link #delete()} explicitly.
-     */
-    @Override
-    public void clear() {
-      nativeResources.detach(this);
-      super.clear();
+  void attach(Pointer... resources) {
+    checkSession();
+    for (Pointer r : resources) {
+      nativeResources.attach(r);
     }
-
-    /** Releases all native resources owned by the referred object, now deleted. */
-    abstract void delete();
-
-    private final NativeResourceCollector nativeResources;
   }
 
-  /**
-   * Collects native references attached to this session and releases their resources if they are no
-   * longer needed.
-   */
-  private static class NativeResourceCollector {
-
-    NativeResourceCollector(ReferenceQueue<Object> garbageQueue) {
-      this.garbageQueue = garbageQueue;
-    }
-
-    void attach(NativeReference nativeRef) {
-      synchronized (nativeRefs) {
-        nativeRefs.put(nativeRef, null);
-      }
-    }
-
-    void detach(NativeReference nativeRef) {
-      synchronized (nativeRefs) {
-        nativeRefs.remove(nativeRef);
-      }
-    }
-
-    void delete(NativeReference nativeRef) {
-      synchronized (nativeRefs) {
-        if (!nativeRefs.keySet().remove(nativeRef)) {
-          return; // safety check
-        }
-      }
-      nativeRef.delete();
-    }
-
-    void deleteAll() {
-      synchronized (nativeRefs) {
-        for (NativeReference nativeRef : nativeRefs.keySet()) {
-          nativeRef.delete();
-        }
-        nativeRefs.clear();
-      }
-    }
-
-    void tryCleanup() {
-      Reference<?> nativeRef;
-      synchronized (nativeRefs) {
-        while ((nativeRef = garbageQueue.poll()) != null) {
-          delete((NativeReference) nativeRef);
-        }
-      }
-    }
-
-    synchronized void startCleanupThread() {
-      if (cleanupInBackground) {
-        return; // ignore if cleanup thread is already running
-      }
-      try {
-        cleanupInBackground = true;
-        cleanupService.execute(
-            new Runnable() {
-              @Override
-              public void run() {
-                try {
-                  while (cleanupInBackground) {
-                    NativeReference nativeRef = (NativeReference) garbageQueue.remove();
-                    delete(nativeRef);
-                  }
-                } catch (InterruptedException e) {
-                  // exit
-                }
-              }
-            });
-      } catch (Exception e) {
-        cleanupInBackground = false;
-        throw e;
+  void detach(Pointer... resources) {
+    checkSession();
+    for (Pointer r : resources) {
+      if (resourceCleanupStrategy == ResourceCleanupStrategy.ON_SAFE_POINTS) {
+          nativeRefs.attach(r);
```

Comment:
What do we need both `nativeRefs` and `nativeResources`? If I understand correctly, `nativeRefs` are detached resources that are cleared only on safe points?

The purpose of safe points was mainly to allow releasing resources while a session is active but without having a separate thread for garbage-collection. Is the GC thread in JavaCPP is always running? If it cannot be disabled, then we don't need to support safe-points. If it can be disabled, can we link this option with the picked-up strategy then?

Reply:
We can remove those options. I would vote for this. I wasn't sure myself what was their purpose...

JavaCPP starts a thread for GC by default, yes, but it should be considered more or less like a safety net. It can be disabled with a system property, but it's pretty much for debug purposes only:
http://bytedeco.org/javacpp/apidocs/org/bytedeco/javacpp/Pointer.html#referenceQueue
I haven't encountered a user who wanted to have it disabled as a "feature", yet anyway.

Reply:
Why do the queues in JavaCPP call `System.gc()`? It's not guaranteed that that method does anything, and what it does is probably GC algorithm dependent. Does this work across all the GCs in Hotspot?

Reply:
Like I said, it's a safety net, precisely because Java does not provide any guarantees that anything gets done at all, it shouldn't be relied upon.

Reply:
Then for the sake of simplicity, I opt to remove these options as well, the timing is still good for doing such changes. Anyway we can't disable the GC thread only at the eager session level anymore so they don't make much sense to have them there, if they have to be restored, it would be in a different place.

Reply:
"Safety net" isn't the right analogy. System.gc() is actually a last-ditch effort, but it usually helps avoid OutOfMemoryError, so that's why it's there.